### PR TITLE
Sort self-replies to the top of descendants

### LINKED ISSUE #6463: [Feature request] If there are replies-to-self in a thread, group them all at the top
Behold:

screen shot 2018-02-11 at 20 08 42

If someone replies before I'm done with my thread it makes the thread hard to read. I think it would be cool if all the replies-to-self could be grouped together at the top, even if some people replied before all the posts were posted, just for continuity of a thread.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### LINKED ISSUE #9315: Wrong order in threads due to regression
Regression in #9296 randomizes the order of statuses in a thread.

### PR DESCRIPTION
**Pull request by Gargron**
Fix #6463

### CODE REVIEW

**Comment by ThibG** [comment]
What happens with the following thread (where “a → b” denotes that b replied to a)?
`alice (1) → alice (2) → bob (3) → alice (4) → alice (5)`

I'm afraid this gives `alice (1) → alice (2) → alice (5) →bob (3) → alice (4)` which would be extremely confusing?

**Comment by Gargron** [comment]
@ThibG That shouldn't happen, because in_reply_to_account_id carries over between self-replies. So if I reply to Bob, then reply to my reply to Bob, in_reply_to_account_id is Bob, not me.

File: app/models/concerns/status_threading_concern.rb [hunk 1]
```diff
@@ -88,7 +88,26 @@ def find_statuses_from_tree_path(ids, account)
...
+      promote_by!(statuses) { |status| status.in_reply_to_account_id == status.account_id }
+    else
+      statuses
+    end
+  end
+
+  def promote_by!(arr, &condition)
+    insert_at = arr.find_index { |item| !condition.call(item) }
+
+    return arr if insert_at.nil?
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/models/concerns/status_threading_concern.rb]: Descendants does not contain the status being replied to originally, as far as I remember. So these are all replies. However, if it's all self-replies, then there is nothing to do here, therefore return array.

[the remark this answers is not in the document]
Comment (ThibG) [reply app/models/concerns/status_threading_concern.rb]: Oh, indeed.

**Review by ClearlyClaire — APPROVED** [review]
I would probably add a spec to cover the case I was confused about (but seems to be handled properly), but otherwise it looks good to me.

**Pull request merged**